### PR TITLE
Support weekofyear pushdown to tiflash

### LINKED ISSUE #4677: Support `weekofyear` in TiFlash
## Feature Request

**Is your feature request related to a problem? Please describe:**

**Describe the feature you'd like:**

Support `weekofyear` in TiFlash
MySQL doc:
https://dev.mysql.com/doc/refman/8.0/en/date-and-time-functions.html#function_weekofyear
TiDB implementation:
https://github.com/pingcap/tidb/blob/5889d4c4e9357a515c472fcf12bb3624fbb00f9e/expression/builtin_time.go#L1502
**Describe alternatives you've considered:**

**Teachability, Documentation, Adoption, Migration Strategy:**

### PR DESCRIPTION
**Pull request by gengliqi**
### What problem does this PR solve?

Issue Number: close #4677

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
Support weekofyear pushdown to tiflash
```

### CODE REVIEW

File: dbms/src/Functions/FunctionsDateTime.h [hunk 1]
```diff
@@ -3238,6 +3238,44 @@ struct TiDBDayOfYearTransformerImpl
...
+    static constexpr auto name = "tidbWeekOfYear";
+
+    static void execute(const Context & context,
+                        const ColumnVector<DataTypeMyTimeBase::FieldType>::Container & vec_from,
+                        typename ColumnVector<ToFieldType>::Container & vec_to,
+                        typename ColumnVector<UInt8>::Container & vec_null_map)
+    {
+        for (size_t i = 0; i < vec_from.size(); ++i)
+        {
+            bool is_null = false;
```
Comment (ywqzzy) [inline dbms/src/Functions/FunctionsDateTime.h:3254]: maybe declare the is_null outside of the for loop？

File: dbms/src/Functions/tests/gtest_weekofyear.cpp [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+{
+class TestWeekOfYear : public DB::tests::FunctionTest
+{
+};
+
+TEST_F(TestWeekOfYear, WeekOfYear)
+try
+{
+    DAGContext * dag_context = context.getDAGContext();
+    UInt64 ori_flags = dag_context->getFlags();
```
Comment (ywqzzy) [inline dbms/src/Functions/tests/gtest_weekofyear.cpp:33]: Maybe move the flag related operations to TestWeekOfYear's setup and  Teardown methods.

Comment (gengliqi) [reply dbms/src/Functions/tests/gtest_weekofyear.cpp:33]: Emmm, there is only one test and the flag is specifically used for testing zero date.

File: dbms/src/Functions/FunctionsDateTime.h [hunk 3]
```diff
@@ -3238,6 +3238,45 @@ struct TiDBDayOfYearTransformerImpl
...
+        if (val.month == 0 || val.day == 0)
+        {
+            context.getDAGContext()->handleInvalidTime(
+                fmt::format("Invalid time value: month({}) or day({}) is zero", val.month, val.day),
+                Errors::Types::WrongValue);
+            is_null = true;
+            return 0;
+        }
+        /// Behavior differences from TiDB:
+        /// for date in ['0000-01-02'), weekofyear is the same with MySQL, while TiDB is offset by one day
```
Comment (guo-shaoge) [inline dbms/src/Functions/FunctionsDateTime.h:3274]:
```suggestion
        /// for '0000-01-02', weekofyear is the same with MySQL, while TiDB is offset by one day
```

**Review by yibin87 — APPROVED** [review]
LGTM

**Review by guo-shaoge — APPROVED** [review]

**Pull request merged**